### PR TITLE
8320886: Unsafe_SetMemory0 is not guarded

### PR DESCRIPTION
**Pull request by JornVernee**
See JBS issue.

Guard the memory access done in Unsafe_SetMemory0 to prevent a SIGBUS error from crashing the VM when a truncated memory mapped file is accessed.

Testing: local `InternalErrorTest`, Tier 1-5

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320886](https://bugs.openjdk.org/browse/JDK-8320886): Unsafe_SetMemory0 is not guarded (**Bug** - P2)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [e9a5247e](https://git.openjdk.org/jdk/pull/16848/files/e9a5247ee00571b4adc4c5c7e627422abcacec42)
 * [Frederic Parain](https://openjdk.org/census#fparain) (@fparain - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16848/head:pull/16848` \
`$ git checkout pull/16848`

Update a local copy of the PR: \
`$ git checkout pull/16848` \
`$ git pull https://git.openjdk.org/jdk.git pull/16848/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16848`

View PR using the GUI difftool: \
`$ git pr show -t 16848`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16848.diff">https://git.openjdk.org/jdk/pull/16848.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16848#issuecomment-1830496051)

### CODE REVIEW

**Comment by franz1981** [comment]
@JornVernee please check what I have done for https://github.com/openjdk/jdk/pull/16760

Maybe there is something we can pull in already on your PR here re it

**Comment by JornVernee** [comment]
@franz1981 I can not interact with PRs that have the OCA label. Once you have signed the OCA, and it is verified (label is removed), I can take a look.

**Comment by franz1981** [comment]
@JornVernee Fair point, in theory I work for Red Hair so it should be sorted out, let me ask @theRealAph about it

**Comment by JornVernee** [comment]
@franz1981 No problem. I wasn't sure whether you worked for Red Hat or not. In that case, it should be enough to post a comment with `/covered` and then one of the community managers will create an entry for your github user name in the OCA database: https://oca.opensource.oracle.com/?ojr=contrib-list

I will stick to the policy for waiting until the OCA label is remove though :)

**Comment by theRealAph** [comment]
> @franz1981 I can not interact with PRs that have the OCA label. Once you have signed the OCA, and it is verified (label is removed), I can take a look.

I can tell you that @franz1981 is a Red Hat employee, and we have signed the OCA for everyone here.

**Comment by JornVernee** [comment]
https://github.com/openjdk/jdk/pull/16760 Looks more performance focused, so I think that should be addressed separately.

**Comment by JornVernee** [comment]
I'm looking into the crash on mac

**Comment by dholmes-ora** [comment]
@JornVernee is there some "new" usage of this method such that it needs guarding? I find it interesting that in the 3+ months that the fix for [JDK-8191278](https://bugs.openjdk.org/browse/JDK-8191278) was developed and reviewed, the need to include `SetMemory0` was never raised once. And that was over 4 years ago.

**Comment by JornVernee** [comment]
> @JornVernee is there some "new" usage of this method such that it needs guarding?

This function is now exposed (indirectly) through the new `MemorySegment::fill` API. There are some rare other uses of `Unsafe.setMemory` in the JDK, and some seem to be operating on user-supplied buffers, so this issue might have existed for longer already. For instance in `GaloisCounterMode`, but it's hard to tell for me where the ByteBuffer that this is operating on comes from.

I went through the original review thread starting here: https://mail.openjdk.org/pipermail/hotspot-dev/2019-February/037058.html but I don't see a reference to `setMemory`.

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 1]
```diff
@@ -969,7 +970,7 @@ address Assembler::locate_operand(address inst, WhichOperand which) {
       assert(which == call32_operand, "jcc has no disp32 or imm");
       return ip;
     default:
-      ShouldNotReachHere();
+      fatal("not handled: 0x0F%2X", 0xFF & *(ip-1));
```
Comment (JornVernee) [inline src/hotspot/cpu/x86/assembler_x86.cpp:973]: I've improved the error reporting a bit here to print out the problematic instruction. I was only able to reproduce the crash through GHA.

**Review by dholmes-ora — APPROVED** [review]
Okay fix seems fine.

Thanks

**Review by fparain — APPROVED** [review]

**Pull request closed without merging**